currency-rates/internal/services/transaction.go

```diff
@@ -0,0 +1,62 @@
+package services
+
+import (
+	"context"
+
+	"github.com/RinaDish/currency-rates/tools"
+)
+
+type DB interface {
+	SetEmail(ctx context.Context, email string) error
+	DeactivateEmail(ctx context.Context, email string) error
+	SetUser(ctx context.Context, email string) error
+	DeleteUser(ctx context.Context, email string) error
+}
+
+type Transaction struct {
+	customerService     CustomerService
+	subscriptionService SubscriptionService
+	logger              tools.Logger
+}
+
+func NewTransactionService(logger tools.Logger, customerService CustomerService, subscriptionService SubscriptionService) Transaction {
+	return Transaction{
+		logger:              logger,
+		subscriptionService: subscriptionService,
+		customerService:     customerService,
+	}
+}
+
+func (transaction Transaction) ExecuteSubscription(ctx context.Context, email string) error {
+	transaction.logger.Info("Starting SAGA transaction")
+
+	
+	if err := transaction.createUser(ctx, email); err != nil {
+		return err
+	}
+
+	
+	if err := transaction.createSubscription(ctx, email); err != nil {
+		err = transaction.compensateCreateUser(ctx, email)
+		return err
+	}
+
+	transaction.logger.Info("SAGA transaction completed successfully")
+
+	return nil
+}
```

Comment:
So first of all SAGA is introduced for distributed systems and firstly for databases per service. Ideally here you could just use normal SQL transaction without overhead =)
But for simplicity let's left as it is